digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -311,124 +311,42 @@ class NodesAndJunctionsService(roadwayDAO: RoadwayDAO, roadwayPointDAO: RoadwayP
-      val rwPoint: Long = getRoadwayPoint(link.roadwayNumber, link.endAddrMValue, username)
-      val linkJunctionPoint = junctionPointDAO.fetchByRoadwayPoint(link.roadwayNumber, link.endAddrMValue, BeforeAfter.Before)
-      if (linkJunctionPoint.isEmpty) {
-        logger.info(s"Creating JunctionPoint with roadwayNumber : ${link.roadwayNumber} addrM: ${link.endAddrMValue} beforeAfter: ${BeforeAfter.Before.value}, junctionId: $junctionId")
-        junctionPointDAO.create(Seq(JunctionPoint(NewIdValue, BeforeAfter.Before, rwPoint, junctionId, None, None, DateTime.now, None, username, Some(DateTime.now), link.roadwayNumber, link.endAddrMValue, link.roadNumber, link.roadPartNumber, link.track, link.discontinuity)))
-        CalibrationPointsUtils.createCalibrationPointIfNeeded(rwPoint, link.linkId, CalibrationPointLocation.EndOfLink, CalibrationPointType.JunctionPointCP, username)
+  def handleJunctionTemplates(roadwayChanges: List[ProjectRoadwayChange], projectLinks: Seq[ProjectLink], mappedRoadwayNumbers: Seq[ProjectRoadLinkChange], username: String = "-"): Unit = {
+    def getRoadwayPointId(roadwayNumber: Long, address: Long, username: String): Long = {
+      val existingRoadwayPoint = roadwayPointDAO.fetch(roadwayNumber, address)
+      val rwPoint = if (existingRoadwayPoint.nonEmpty) {
+        existingRoadwayPoint.get.id
+      } else {
+        roadwayPointDAO.create(roadwayNumber, address, username)
+      }
+      rwPoint
+    }
+
+    def createJunctionPointIfNeeded(r: BaseRoadAddress, pos: BeforeAfter, addr: Long, roadsTo: Seq[BaseRoadAddress], roadsFrom: Seq[BaseRoadAddress],
+                                    newJunctionId: Option[Long] = Option.empty[Long]): Unit = {
+      val roadwayPointId = getRoadwayPointId(r.roadwayNumber, addr, username)
+      val junctionPoint = junctionPointDAO.fetchByRoadwayPoint(r.roadwayNumber, addr, pos)
+      if (junctionPoint.isEmpty) {
+        val junctionId = newJunctionId match {
+          case Some(junctionId) => if (junctionId == NewIdValue) {
+            logger.info(s"Creating Junction for roadwayNumber: ${r.roadwayNumber}, beforeAfter: ${pos.value}, addrM: $addr")
+            junctionDAO.create(Seq(Junction(NewIdValue, None, None, r.startDate.get, None, DateTime.now, None, username, Some(DateTime.now)))).head
+          } else junctionId
+          case None =>
+            val junctionInRoadAddress = junctionPointDAO.fetchByRoadwayPoint(r.roadwayNumber, addr, pos)
+            val junctionInRoadsToRoadAddress = roadsTo.flatMap(r => junctionPointDAO.fetchByRoadwayPoint(r.roadwayNumber, r.endAddrMValue, BeforeAfter.Before))
+            val junctionInRoadsFromRoadAddress = roadsFrom.flatMap(r => junctionPointDAO.fetchByRoadwayPoint(r.roadwayNumber, r.startAddrMValue, BeforeAfter.After))
+            (junctionInRoadAddress ++ junctionInRoadsToRoadAddress ++ junctionInRoadsFromRoadAddress).map(_.junctionId).toSeq.distinct.headOption match {
```

Comment:
since we use distinct here we could do some getOrElse checks so if we find something in first fetch, we dont need to do the others below